docs/development/pigs/pig-007.rst

```diff
@@ -0,0 +1,409 @@
+of "forward folded" models. These take a fixed data grid, typically defined by the
+exposure map or effective area table, on which the model is evaluated. After application of
+instrument reponse functions, the number of prediced counts is returned. This can be
+either accessed by a `.npred` property or is returned by `.evaluate()`. We propose
+to introduce the following classes:
+
+`SpectralIRFModel`
+~~~~~~~~~~~~~~~~~~
+A "forward folded" spectral model, that applies energy dispersion and effective
+area to a `SpectralModel`. It corresponds in functionality to the current `CountsPredictor`.
+
+.. code-block::
+
+    from gammapy.spectrum import Powerlaw, SpectralIRFModel
+
+    pwl = Powerlaw()
+    spectral_irf_model = SpectralIRFModel(model=pwl, aeff=aeff, livetime=livetime,
+                                          edisp=edisp)
+    npred = spectral_irf_model.npred
+    dnde = spectral_irf_model.dnde
+    flux = spectral_irf_model.flux
+
+
+`(SpatialIRFModel)`
+~~~~~~~~~~~~~~~~~~~
+A convolved model, that applies the PSF and exposure to a `SpatialModel`. The
+need for this model is not clear, as we might solve the use case of morphology
+fitting by a combined spatial and spectral analysis with one energy bin.
+
+`SourceIRFModel`
+~~~~~~~~~~~~~~~~
+A "forward folded" source model, that applies IRFs to a `SourceModel`
+instance and returns an integrated quantity corresponding to predicted counts. It
+can only be evaluated on a fixed grid, which is defined on initialization by the
+exposure map. In addition it takes reduced PSF and EDISP objects. In functionality
+it corresponds to the current `MapEvaluator`, but with the model parameters attached
+and the background handling removed. Additional "hidden" IRF parameters e.g. to propagate
+systematics could be optionally introduced later. This class will replace the current
+`MapEvaluator`.
+
+.. code-block::
+
+    from gammapy.cube import SourceIRFModel, SourceModel
+
+    model = SourceModel()
+    source_irf_model = SourceIRFModel(model, exposure=exposure, edisp=edisp, psf=psf)
+    npred = source_irf_model.npred
+    dnde = source_irf_model.dnde
+    flux = source_irf_model.flux
+
+
+Improve `SourceModels` class
+----------------------------
+The existing `CompoundSkyModel` is a nice very generic abstraction to support
+any kind of arithmetics between `SkyModel` objects, but the number of use
+cases for other operators, except for "+" is very limited and
+can always be achieved by implementing a custom model. I/O and component handling
+of this hierarchical model component structure is intrinsically difficult. For
+this reason we propose to first support and improve the existing `SourceModels` that
+implements an easier to handle flat hierarchy for model components. Support for
+arbitrary model arithmetics can be introduced, if needed, after Gammapy v1.0. We
+propose to remove the `CompoundSourceModel` and reimplement the `+` operator using
+the `SourceModels` class.
```

Comment:
+1 to have `model1 + model2` result in a `SourceModels` object, i.e. a list of two models. I agree that's the much nicer object compared to the compound model, where `model1 + model2 + model3 + ...` puts an effectively list of independent objects in a tree data structure that's hard to serialise / repr.

But I would suggest to split out your proposal to remove `CompoundSourceModel` and maybe `SpectralCompoundModel` and to have that discussion / decision a bit separately.

Maybe it would be useful to have some statement on what models we want to be able to serialise and which ones not? I think there are several different kinds of complex models (compounds, probably also IRF? and others) that we will not be able to serialise at all, or to completely serialise? I think ctools can only serialise "flat lists" of independent models, but 3ML / astromodels has a recursive / hierarchical solution for parameter names and models. This PIG should clarify what we plan to develop in Gammapy.